### PR TITLE
[KYUUBI #4937] Cleanup spark catalog shim and renamed to catalog utils

### LINKED ISSUE #4937: Remove SparkCatalogShim in Spark engine
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

`SparkCatalogShim` was introduced as a shim layer for compatibility with Spark 2.4, but it does not happen and has less opportunity to happen in the future, let's remove it to simply the code.

### How should we improve?

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [X] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by bowenliang123**
### _Why are the changes needed?_

to close #4937.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+
+  private def getSchemasWithPattern(catalog: CatalogPlugin, schemaPattern: String): Seq[String] = {
+    val p = schemaPattern.r.pattern
+    listAllNamespaces(catalog).flatMap { ns =>
+      val quoted = ns.map(quoteIfNeeded).mkString(".")
+      if (p.matcher(quoted).matches()) {
+        Some(quoted)
+      } else {
+        None
+      }
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:153]:
```suggestion
      if (p.matcher(quoted).matches()) Some(quoted) else None
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        }
+      case tc: TableCatalog =>
+        val tp = tablePattern.r.pattern
+        val identifiers = namespaces.flatMap { ns =>
+          tc.listTables(ns).filter(i => tp.matcher(quoteIfNeeded(i.name())).matches())
+        }
+        identifiers.map { ident =>
+          // TODO: restore view type for session catalog
+          val comment = if (ignoreTableProperties) ""
+          else tc.loadTable(ident).properties().getOrDefault(TableCatalog.PROP_COMMENT, "")
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:206]: It was reported by dozen users that metadata refresh takes a long time, let's add the comment to the code to make the learner easy to understand.

```suggestion
          val comment = if (ignoreTableProperties) ""
          else { // load table is a time consuming operation
            tc.loadTable(ident).properties().getOrDefault(TableCatalog.PROP_COMMENT, "")
          }
```

Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala]: Adopted.

**Comment by pan3793** [comment]
A related PR may worth to evaluate https://github.com/apache/kyuubi/pull/4465

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+  def getCatalogs(spark: SparkSession): Seq[Row] = {
+
+    // A [[CatalogManager]] is session unique
+    val catalogMgr = spark.sessionState.catalogManager
+    // get the custom v2 session catalog or default spark_catalog
+    val sessionCatalog = invokeAs[AnyRef](catalogMgr, "v2SessionCatalog")
+    val defaultCatalog = catalogMgr.currentCatalog
+
+    val defaults = Seq(sessionCatalog, defaultCatalog).distinct.map(catalog =>
+      DynMethods.builder("name").impl(catalog.getClass).buildChecked(catalog).invoke[String]())
```
Comment (bowenliang123) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:60]: While `ReflectUtils.invokeAs` supports declared method only, I have to use DynMethods for calling method of inherited methods here manually.  cc @pan3793

**Comment by bowenliang123** [comment]
> A related PR may worth to evaluate #4465

Seems not enough evidence for the necessity of the suggested changes in that PR.
As for me, checking the catalog name to `spark_catalog` is more strict than matching in TableCatalog class, and it's appropriate to be placed in the first matching case here.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala (hunk 3 above)
Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:60]: Is it possible to access all methods by an unified reflection api?

Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:60]: Shall I refactor the `ReflectUtils.invokeAs` for this, by calling declared methods first and then the inherited methods? And I will throw corresponding method candidates to either one.

Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:60]: sgtm

**Comment by pan3793** [comment]
> Seems not enough evidence for the necessity of the suggested changes in that PR.

The key is order, there were some issues(see https://github.com/apache/kyuubi/pull/4219) when `DelegatingCatalogExtension` is used as `spark_catalog`.

**Comment by pan3793** [comment]
The change of this PR is already big, it's valuable to do some further changes in follow-up PRs.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala (hunk 3 above)
Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/util/SparkCatalogUtils.scala:60]: OK, let's make it in the  follow-ups.

**Review by pan3793 — APPROVED** [review]

**Comment by bowenliang123** [comment]
Thanks, merged to master.
